Document/0x07a-Testing-Authentication-and-Session-Management.md

```diff
@@ -462,100 +462,71 @@ Most of the frameworks have a parameter to configure the session timeout. This p
-
--- TODO [Confirm remark on "Use the &lt;sup&gt; tag to reference external sources, e.g. Meyer's recipe for tomato soup<sup>[1]</sup>."] --
-
--- TODO [Develop content on Testing 2-Factor Authentication with source code] --
-
-
-#### Dynamic Analysis
-
--- TODO [Describe how to test for this issue "Testing 2-Factor Authentication" by running and interacting with the app. This can include everything from simply monitoring network traffic or aspects of the app’s behavior to code injection, debugging, instrumentation, etc.] --
-
-#### Remediation
+Applications that offer access to sensitive data or critical functions, might require users additionally to re-authenticate with a stronger authentication mechanism. For example, after logging in via biometric authentication (e.g. TouchID) into a banking app, a user might need to do a so called "Step-up Authentication" again through OTP in order to execute a bank transfer.
 
--- TODO [Describe the best practices that developers should follow to prevent this issue "Testing 2-Factor Authentication".] --
+A key advantage of step-up authentication is improved usability for the user. A user is asked to authenticate with
+the additional factor only when necessary.
 
-#### References
-
-##### OWASP Mobile Top 10 2016
-* M4 - Insecure Authentication - https://www.owasp.org/index.php/Mobile_Top_10_2016-M4-Insecure_Authentication
-
-##### OWASP MASVS
-* 4.8: "A second factor of authentication exists at the remote endpoint and the 2FA requirement is consistently enforced."
-
-##### CWE
-
--- TODO [Add relevant CWE for "Testing 2-Factor Authentication"] --
-- CWE-312 - Cleartext Storage of Sensitive Information
-
-##### Info
-
-- [1] Meyer's Recipe for Tomato Soup - http://www.finecooking.com/recipes/meyers-classic-tomato-soup.aspx
-- [2] Another Informational Article - http://www.securityfans.com/informational_article.html
-
-##### Tools
-
--- TODO [Add relevant tools for "Testing 2-Factor Authentication"] --
-* Enjarify - https://github.com/google/enjarify
 
+#### Static Analysis
 
+When server-side source code is available, first identify how a second factor or step-up authentication is used and enforced. Afterwards locate all endpoints with sensitive and privileged information and functions: they are the ones that need to be protected. Prior to accessing any item, the application must make sure the user has already passed 2FA or the step-up authentication and that he is allowed to access the endpoint.
 
+2FA or step-up authentication shouldn't be implemented from scratch, instead they should be build on top of available libraries that offer this functionality. The libraries used on the client and server side should be identified and the usage of the available APIs/functions should be verified if they are used accordingly to best practices. Widely used OTP mechanisms are for example:
 
-### Testing Step-up Authentication
+- Google Authenticator
+- Microsoft Authenticator
+- Authy
```

Comment:
Aren't these all separate applications that run on the mobile device? Can hou provide a few different examples as well? Because an argument that might be used by the business is that it is not convenient to switch from one app to another.